### PR TITLE
Add GitHub actions to address GNU/Clang warnings

### PR DESCRIPTION
**Pull request by juan-lunarg**
The intent here is to address warnings coming from `include/vk_mem_alloc.h` and make sure it stays fixed by having it as part of CI.

NOTE: By default VMA_BUILD_ERROR is OFF to not cause package managers any trouble.

Split up into atomic commits to highlight fixing the compiler warnings.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -8,6 +8,24 @@ add_library(VulkanMemoryAllocator
...
+endif()
+
+if(${CMAKE_CXX_COMPILER_ID} MATCHES "(GNU|Clang)")
+    target_compile_options(VulkanMemoryAllocator PRIVATE
+        -Wpedantic
+        -Wall
+        -Wextra
+        -Wno-unused-parameter
+        -Wno-unused-variable
+        -Wno-missing-field-initializers
```
Comment (juan-lunarg) [inline src/CMakeLists.txt:25]: My goal is to fix the remaining warnings in another PR. I don't want to complicate this one too much.

**Comment by juan-lunarg** [comment]
I'd like to add some rational to this.

While some users consume VMA via system include commands like `-isystem dir` not all do. And many users don't even know about `-isystem dir`. So they just silence warnings in their project after introducing a dependency.

And for MVSC the problem is even worse because until very recently they didn't even have an equivalent to `-isystem`.

So the best thing for project to do is to make sure that their public interface is as clean as possible. Which is what I'm trying to do here.

**Comment by adam-sawicki-a** [comment]
Thank you for this proposal.

Successful build of the library is already validated in Windows + Visual Studio:
https://ci.appveyor.com/project/adam-sawicki-amd/vulkanmemoryallocator/branch/master
As well as Linux + GCC and Clang:
https://app.travis-ci.com/github/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator

We may consider enabling GitHub actions instead in the future.

Regarding compiler warnings, documentation explicitly says in chapter "Features not supported":

> Code free of any compiler warnings. Maintaining the library to compile and work correctly on so many different platforms is hard enough. Being free of any warnings, on any version of any compiler, is simply not feasible. There are many preprocessor macros that make some variables unused, function parameters unreferenced, or conditional expressions constant in some configurations. The code of this library should not be bigger or more complicated just to silence these warnings. It is recommended to disable such warnings instead.

The change you proposed that adds `break;` in two places actually changes code logic and makes it incorrect. Fallthrough is intentional there.

**Comment by turol** [comment]
> The change you proposed that adds `break;` in two places actually changes code logic and makes it incorrect. Fallthrough is intentional there.

You should add a "fallthrough" comment to those. It both tells the reader that it's supposed to fall through and silences the warning.

**Comment by turol** [comment]
Nope, GCC and Clang (I think) will interpret a comment containing the word "fallthrough" the same and shut up about it.

**Comment by adam-sawicki-a** [comment]
I added macro `VMA_FALLTHROUGH` non-empty only when C++17 is enabled + a comment with word "Fallthrough". Please let me know if it works for you.

**Comment by juan-lunarg** [comment]
That handles it nicely :+1:

**Pull request closed without merging**